Repeated use of inline queue breaks amalgamated build

## Problem description

When building the current master code with `amalgamation=yes`, redefinition of the `GtInl_Queueelem` type before including `match/queue-inline.h` triggers a compile error:

```
[compile amalgamation.o]
In file included from obj/amalgamation.c:542:0:
/home/travis/build/genometools/genometools/src/match/sfx-diffcov.c:105:27: error: conflicting types for â€˜GtInl_Queueelemâ€™
 typedef GtDcPairsuffixptr GtInl_Queueelem;
                           ^
In file included from obj/amalgamation.c:116:0:
/home/travis/build/genometools/genometools/src/core/sequence_buffer_dust.c:26:23: note: previous declaration of â€˜GtInl_Queueelemâ€™ was here
 typedef unsigned char GtInl_Queueelem;
                       ^
In file included from obj/amalgamation.c:542:0:
/home/travis/build/genometools/genometools/src/match/sfx-diffcov.c: In function â€˜dc_processunsortedrangeâ€™:
/home/travis/build/genometools/genometools/src/match/sfx-diffcov.c:815:43: error: incompatible type for argument 2 of â€˜gt_inl_queue_addâ€™
   gt_inl_queue_add(dcov->rangestobesorted,pairelem,false);
                                           ^
In file included from /home/travis/build/genometools/genometools/src/core/sequence_buffer_dust.c:27:0,
                 from obj/amalgamation.c:116:
/home/travis/build/genometools/genometools/src/match/queue-inline.h:123:20: note: expected â€˜GtInl_Queueelem {aka unsigned char}â€™ but argument is of type â€˜GtDcPairsuffixptr {aka struct <anonymous>}â€™
 static inline void gt_inl_queue_add(GtInl_Queue *q, GtInl_Queueelem elem,
                    ^
In file included from obj/amalgamation.c:542:0:
/home/travis/build/genometools/genometools/src/match/sfx-diffcov.c: In function â€˜dc_sortremainingsamplesâ€™:
/home/travis/build/genometools/genometools/src/match/sfx-diffcov.c:1026:14: error: incompatible types when assigning to type â€˜GtDcPairsuffixptr {aka struct <anonymous>}â€™ from type â€˜GtInl_Queueelem {aka unsigned char}â€™
     thispair = gt_inl_queue_get(dcov->rangestobesorted);
              ^
Makefile:840: recipe for target 'obj/obj/amalgamation.o' failed
``` 

## Exact command line call triggering the problem

In principle:
```
make amalgamation=yes
```

See

## Example minimal input triggering the problem

Git master e5c5d6ef9b4305e6238924914e16a6a55a77a6d6

## What GenomeTools version are you reporting an issue for (as output by `gt -version`)?

git master

## Did you compile GenomeTools from source? If so, please state the `make` parameters used.

See above

## What operating system (e.g. Ubuntu, Mac OS X), OS version (e.g. 15.10, 10.11) and platform (e.g. x86_64) are you using?

Ubuntu 16.04.6 LTS x86_64
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609

--- a/scripts/skmake.rb
+++ b/scripts/skmake.rb
@@ -96,7 +96,11 @@ def makecompilerflags(fp,options)
   else
     fp.print " cairo=no"
   end
-  fp.print " #{extracpp} popcnt=yes CC='ccache " + ENV["CC"] + "'"
+  compiler="cc"
+  if ENV.has_key?("CC")
+    compiler=ENV["CC"]
+  end
+  fp.print " #{extracpp} popcnt=yes CC='ccache " + compiler + "'"
   if not options.fileargs.nil?
     filenames=options.fileargs.join(" ")
     fp.puts " #{filenames}"

--- a/src/ltr/ltrdigest_strand_assign_visitor.c
+++ b/src/ltr/ltrdigest_strand_assign_visitor.c
@@ -40,7 +40,7 @@ const GtNodeVisitorClass* gt_ltrdigest_strand_assign_visitor_class(void);
 
 static int gt_ltrdigest_strand_assign_visitor_feature_node(GtNodeVisitor *nv,
                                                  GtFeatureNode *fn,
-                                                 GtError *err)
+                                                 GT_UNUSED GtError *err)
 {
   GtLTRdigestStrandAssignVisitor *lv;
   GtFeatureNodeIterator *fni;

--- a/src/match/seed-extend.c
+++ b/src/match/seed-extend.c
@@ -1322,7 +1322,7 @@ static void gt_rf_seed_extend_prettyprint(bool forxdrop,void *info)
   GtProcessinfo_and_querymatchspaceptr *info_querymatch
     = (GtProcessinfo_and_querymatchspaceptr *) info;
   GtUword userdefinedleastlength, errorpercentage;
-  double evalue, bit_score, evalue_threshold;
+  double evalue = 0.0, bit_score = 0.0, evalue_threshold = 0.0;
 
   gt_assert(info_querymatch != NULL);
   if (forxdrop)

--- a/src/match/sfx-diffcov.c
+++ b/src/match/sfx-diffcov.c
@@ -52,9 +52,11 @@
   indexes in the sorted array of suffixes.
 
 */
-
+#include <stdio.h>
+#include <stdlib.h>
 #include <stdbool.h>
 #include <math.h>
+#include "core/types_api.h"
 #include "core/arraydef_api.h"
 #include "core/assert_api.h"
 #include "core/error_api.h"
@@ -91,7 +93,7 @@ typedef unsigned short Diffvalue;
 typedef struct
 {
   GtUword key,
-                suffixstart;
+          suffixstart;
 } GtDcItventry;
 
 typedef struct
@@ -102,9 +104,143 @@ typedef struct
 
 GT_DECLAREARRAYSTRUCT(GtDcPairsuffixptr);
 
-typedef GtDcPairsuffixptr GtInl_Queueelem;
+typedef GtDcPairsuffixptr GtInlDCQueueelem;
 
-#include "queue-inline.h"
+typedef struct
+{
+ GtInlDCQueueelem *queuespace;  /* the space to store the queue elements */
+ GtUword enqueueindex, /* entry into which element is to be enqued */
+         dequeueindex, /* last element of queue */
+         queuesize,    /* size of the queue */
+         noofelements; /* no ofelements between enqueueindex+1
+                          and dequeindex */
+} GtInlDCQueue;
+
+typedef int (*GtInlDCQueueprocessor)(GtInlDCQueueelem *,void *info);
+
+/*
+  The following function delivers an empty queue with a reservoir of
+  \texttt{size} elements to be stored in the queue. The
+  reservoir can, if necessary, be enlarged.
+*/
+
+static GtInlDCQueue *gt_inl_dc_queue_new(GtUword queuesize)
+{
+  GtInlDCQueue *q;
+
+  q = gt_malloc(sizeof (*q));
+  gt_assert(queuesize > 0);
+  q->queuespace = gt_malloc(sizeof (*q->queuespace) * queuesize);
+  q->noofelements = 0;
+  q->queuesize = queuesize;
+  q->dequeueindex = q->enqueueindex = queuesize - 1;
+  return q;
+}
+
+/*
+  The following function frees the space required for the queue.
+*/
+
+static void gt_inl_dc_queue_delete(GtInlDCQueue *q)
+{
+  if (q != NULL)
+  {
+    gt_free(q->queuespace);
+    gt_free(q);
+  }
+}
+
+/*
+  The following function returns true iff the queue is empty.
+*/
+
+static bool gt_inl_dc_queue_isempty(const GtInlDCQueue *q)
+{
+  gt_assert(q != NULL);
+  return (q->noofelements == 0) ? true : false;
+}
+
+/*
+  The following function resizes the queue by doubling the
+  space reservoir.
+*/
+
+static void gt_inl_dc_queue_extend(GtInlDCQueue *q,bool doublesize)
+{
+  GtUword addconst, idx, newsize;
+
+  gt_assert(q != NULL);
+  if (doublesize)
+  {
+    addconst = q->queuesize;
+  } else
+  {
+    addconst = GT_MIN(1024UL,q->queuesize);
+  }
+  newsize = q->queuesize + addconst;
+  q->queuespace = gt_realloc(q->queuespace,sizeof (*q->queuespace) * newsize);
+  gt_assert(q->enqueueindex == q->dequeueindex);
+  gt_assert(addconst > 0);
+  for (idx=q->queuesize-1; idx>q->enqueueindex; idx--)
+  {
+    q->queuespace[idx+addconst] = q->queuespace[idx];
+  }
+  q->enqueueindex += addconst;
+  /*
+  printf("from queue of size "GT_WU" to queue of size "GT_WU"\n",q->queuesize,
+         newsize);
+  printf("now enqueindex="GT_WU",dequeuindex="GT_WU"\n",
+         q->enqueueindex,q->dequeueindex);
+  */
+  q->queuesize = newsize;
+}
+
+/*
+  The following function adds an element \texttt{elem} to the end of
+  the queue.
+*/
+
+static void gt_inl_dc_queue_add(GtInlDCQueue *q, GtInlDCQueueelem elem,
+                                    bool doublesize)
+{
+  gt_assert(q != NULL);
+  if (q->noofelements == q->queuesize)
+  {
+    gt_inl_dc_queue_extend(q,doublesize);
+  }
+  q->noofelements++;
+  q->queuespace[q->enqueueindex] = elem;
+  if (q->enqueueindex > 0)
+  {
+    q->enqueueindex--;
+  } else
+  {
+    q->enqueueindex = q->queuesize - 1; /* dequeuindex < queuesize-1 */
+  }
+}
+
+/*
+  The following function removes the element \texttt{elem} from the
+  start of the queue.
+*/
+
+static GtInlDCQueueelem gt_inl_dc_queue_get(GtInlDCQueue *q)
+{
+  GtInlDCQueueelem value;
+
+  gt_assert(q != NULL && q->noofelements > 0);
+  q->noofelements--;
+  value = q->queuespace[q->dequeueindex];
+  if (q->dequeueindex > 0)
+  {
+    q->dequeueindex--;
+  } else
+  {
+    q->dequeueindex = q->queuesize - 1;  /* != enqueueindex, since at least
+                                            one elem */
+  }
+  return value;
+}
 
 typedef struct
 {
@@ -156,7 +292,7 @@ struct GtDifferencecover
                 currentdepth;
   GtCodetype maxcode;
   GtDcFirstwithnewdepth firstwithnewdepth;
-  GtInl_Queue *rangestobesorted;
+  GtInlDCQueue *rangestobesorted;
   GtDcItventry *itvinfo;
   GtArrayGtDcPairsuffixptr firstgeneration;
   GtLcpvalues *samplelcpvalues;
@@ -812,7 +948,7 @@ static void dc_processunsortedrange(GtDifferencecover *dcov,
   }
   pairelem.blisbl = blisbl;
   pairelem.width = width;
-  gt_inl_queue_add(dcov->rangestobesorted,pairelem,false);
+  gt_inl_dc_queue_add(dcov->rangestobesorted,pairelem,false);
   dcov->currentqueuesize++;
   if (dcov->maxqueuesize < dcov->currentqueuesize)
   {
@@ -1020,10 +1156,10 @@ static void dc_sortremainingsamples(GtDifferencecover *dcov)
     dc_sortsuffixesonthislevel(dcov,pairptr->blisbl, pairptr->width);
   }
   GT_FREEARRAY(&dcov->firstgeneration,GtDcPairsuffixptr);
-  while (!gt_inl_queue_isempty(dcov->rangestobesorted))
+  while (!gt_inl_dc_queue_isempty(dcov->rangestobesorted))
   {
     GtDcPairsuffixptr thispair;
-    thispair = gt_inl_queue_get(dcov->rangestobesorted);
+    thispair = gt_inl_dc_queue_get(dcov->rangestobesorted);
     gt_assert(dcov->currentqueuesize > 0);
     dcov->currentqueuesize--;
     dc_sortsuffixesonthislevel(dcov,thispair.blisbl,thispair.width);
@@ -1031,7 +1167,7 @@ static void dc_sortremainingsamples(GtDifferencecover *dcov)
   gt_logger_log(dcov->logger,"maxqueuesize="GT_WU"",dcov->maxqueuesize);
   gt_free(dcov->itvinfo);
   dcov->itvinfo = NULL;
-  gt_inl_queue_delete(dcov->rangestobesorted);
+  gt_inl_dc_queue_delete(dcov->rangestobesorted);
   dcov->rangestobesorted = NULL;
 }
 
@@ -1194,7 +1330,7 @@ static void dc_differencecover_sortsample(GtDifferencecover *dcov,
   dcov->multimappower = gt_bcktab_multimappower(dcov->bcktab);
   dcov->maxcode = gt_bcktab_numofallcodes(dcov->bcktab) - 1;
   esr1 = gt_encseq_create_reader_with_readmode(dcov->encseq,dcov->readmode,0);
-  dcov->rangestobesorted = gt_inl_queue_new(GT_MAX(16UL,
+  dcov->rangestobesorted = gt_inl_dc_queue_new(GT_MAX(16UL,
                                                    GT_DIV2(dcov->maxcode)));
   dcov->filltable = gt_filllargestchartable(dcov->numofchars,
                                             dcov->prefixlength);
@@ -1481,7 +1617,7 @@ static void dc_differencecover_sortsample0(GtDifferencecover *dcov,
   dcov->bcktab = NULL;
   dcov->multimappower = NULL;
   dcov->maxcode = 0;
-  dcov->rangestobesorted = gt_inl_queue_new(GT_MAX(16UL,
+  dcov->rangestobesorted = gt_inl_dc_queue_new(GT_MAX(16UL,
                                             GT_DIV2(dcov->maxcode)));
   dcov->filltable = NULL;
   dcov->leftborder = NULL;

